src/neighbors.cc

```diff
@@ -616,9 +565,8 @@ neighborsUdpPing(HttpRequest * request,
 
     reqnum = icpSetCacheKey((const cache_key *)entry->key);
 
-    for (i = 0, p = first_ping; i++ < Config.npeers; p = p->next) {
-        if (p == nullptr)
-            p = Config.peers;
+    for (auto peer = CurrentCachePeers().nextPollStart(); peer != CurrentCachePeers().end(); ++peer) {
+        auto p = peer->get();
```

Comment:
AFAICT, this new loop will often miss cache peers. We want to iterate _all_ peers, and not just `[firstPing, end)` peers. Think round-robin from an arbitrary starting point...

The easiest way to implement this will probably result in something like this:

```suggestion
    for (size_t i = 0; i < Config.peers->size(); ++i) {
        const auto p = Config.peers->nextPeerToPoll().get();
```

... and keep the original Config.peers nil check above intact. This approach also avoids ugly const-casting in nextPeerToPoll(). The method will Assure() that at least one peer exists and will always return a peer.

----

With a lot more work, we could make it nicer, but I do _not_ think this improvement is worth our time spent on it:

```suggestion
    for (const auto peer:  Config.peers->peersToPoll()) {
        const auto p = peer.get();
```

Reply:
You are right, I mistakenly thought that the loop should poll only from first_ping (discarding the initial peers). I could not figure out how to adjust this with a single nextPeerToPoll() method - so in the solution I posted there are still two methods: nextPeerToPoll() and nextPollStart(). I have not tested it yet.